fastembed/sparse/splade_pp.py

```diff
@@ -0,0 +1,124 @@
+from typing import Any, Dict, Iterable, List, Optional, Tuple, Union
+
+import numpy as np
+
+from fastembed.common.onnx_model import OnnxModel, EmbeddingWorker
+from fastembed.common.utils import define_cache_dir
+from fastembed.sparse.sparse_embedding_base import SparseEmbedding, SparseTextEmbeddingBase
+
+supported_splade_models = [
+    {
+        "model": "prithvida/SPLADE_PP_en_v1",
+        "vocab_size": 30522,
+        "description": "Independent Implementation of SPLADE++ Model for English",
+        "size_in_GB": 0.532,
+        "sources": {
+            "hf": "nirantk/SPLADE_PP_en_v1",
+        },
+    },
+]
+
+
+class SpladePP(SparseTextEmbeddingBase, OnnxModel[SparseEmbedding]):
+
+    @classmethod
+    def _post_process_onnx_output(cls, output: Tuple[np.ndarray, np.ndarray]) -> Iterable[SparseEmbedding]:
+        logits, attention_mask = output
+        relu_log = np.log(1 + np.maximum(logits, 0))
+
+        weighted_log = relu_log * np.expand_dims(attention_mask, axis=-1)
+
+        max_val = np.max(weighted_log, axis=1)
+
+        # Score matrix of shape (batch_size, vocab_size)
+        # Most of the values are 0, only a few are non-zero
+        scores = np.squeeze(max_val)
+        for row_scores in scores:
```

Comment:
I'm not a huge fan of thse large lists being iterated over. Can we do this in numpy as well please?

Reply:
Yeah, but:

- we iterate over the `batch_size` dimensions, which is not supposed to be very large
- We still benefit from SIMD, cause we apply operations to the whole vocab_size
- result of `nonzero` for the matrix will require iterations on individual elements, as it won't be uniform

So, I couldn't find better alternative. Please feel free to propose something

Reply:
Let me profile something faster and if it's meaningfully faster — this might be the most readable for sure